### ISSUE TITLE
Edge-triggered semantics: change from one screen to another should be properly anounced

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#13704]
When the user navigates from one screen to another that should be properly announced by a sound effect and possibly a user provided announcement text.

See also:
* https://github.com/flutter/flutter/issues/10596
* https://github.com/flutter/flutter/issues/1754
* https://github.com/flutter/flutter/issues/12171

**Comment by goderbauer** [comment]
On Android, we need to produce a `TYPE_WINDOW_STATE_CHANGED` event and set its text to whatever we want announced for the route change (this is usually the title of the new screen).

On iOS, we need to produce a `UIAccessibilityScreenChangedNotification` instead of `UIAccessibilityLayoutChangedNotification`.

**Comment by jonahwilliams** [comment]
I've built a working implementation of this for Android - still working on the iOS version.  To get an idea of how these notifications are used I looked around some existing apps with talkback on, mostly navigation and opening the drawer it seems.  Based on hixie's original suggestion, I include the `routeName` as a part of semantics.

Then on the native side, when traversing the semantics tree I also build a List of the route names from the in-order traversal.  We traverse the list with the cached previous list and announce the first different `routeName`.  This works for what I think are the most common cases, adding or removing a single `routeName` from the semantics tree.  Examples:
```
old [ home ] new [home, navigation]       -> announces navigation
old [home, navigation] new [home]         -> announces home
old [home] -> new [home, foo, bar, baz]   -> announces foo
```

@goderbauer do you know of any other resources to consult for this accessibility work? (or any use cases where this strategy won't work)

**Comment by Hixie** [comment]
can you talk more about the last case? what's a concrete case where this would happen, and does announcing foo make more sense than announcing baz, or maybe foo, bar, and baz?

**Comment by jonahwilliams** [comment]
@Hixie 

In the case where multiple new `routeName`s are found in the semantics tree from a single update, I think the intention is a bit unclear.  We could announce the first (current), the last, or all of them - well probably not all of them since that would be spammy.  I would say that generally, `routeNames` higher in the tree should be considered more important.  Checking the diff to announce the changed child (navigation) is more a special case.  The alternative I think is just too vulnerable to surprises, a great grand child with a mislabeled semantics changing the name of the route for example.

It's mostly theoretical at this point, since I haven't found any existing examples of nested window change notifications.  It could be that in a real app the list of routeNames will never change in this way.

**Comment by goderbauer** [comment]
@jonahwilliams There are not a lot of resources out there that document this. Like you already did, I am mostly reverse engineering this from looking at the adb logs when navigating around existing native apps.

Your proposal sounds good. Do you have a practical example where there would be multiple new routes in the tree? Is the last one always the most specific one? In that case, that should probably be the one announced?

**Comment by goderbauer** [comment]
Also for the multiple routeNames changing: Is there an equivalent for this use case in native android? How does it handle it?

**Comment by jonahwilliams** [comment]
@Hixie , @goderbauer 
Thinking about route names from the perspective of specificity, using the last name makes more sense.  It is also probably much easier to explain.

And no examples of this in a real app.  The closest I can think of is maybe multiple modals which stack and require an announcement of some sort.

**Comment by jonahwilliams** [comment]
I'm going to make one more modification to my existing PRs to add a `isRoute` parameter.  This would allow us to mark something as a semantic route without needing a unique name.  I can handle the diffing by tracking the ids then.  This is probably necessary for cupertino style apps which don't announce (or have) distinct route names.

**Comment by Hixie** [comment]
If we only support reporting this for certain kinds of nodes, we can probably reuse one of the existing string slots (like value) instead of having a new string slot.

**Comment by goderbauer** [comment]
@Hixie Maybe we could have an isRoute flag and if it is set the value is used as the route name? Is that what you mean?

**Comment by Hixie** [comment]
@jonahwilliams suggested adding an isRoute flag; I'm saying if we do that, we should probably reuse "value", instead of introducing "route", as the route name.

I don't have a strong opinion on whether isRoute makes sense. I can see arguments both directions. The main argument against would be that it will typically mean we have to introduce yet another node, which we may not want to do. The main argument in favour is that if you go change route but the new route has the same name, you wouldn't report it unless you're comparing IDs, which might be confusing. (Then again, reporting the same route name again might be confusing too.)

**Comment by jonahwilliams** [comment]
Also on iOS it is not typically to read the route name on transitions, so it would seem to odd to force users working on an iOS version to specify a relatively unique string for that

**Comment by jonahwilliams** [comment]
I'll take a look at whether we could reuse an existing node for the name

**Comment by jonahwilliams** [comment]
Fixed by https://github.com/flutter/flutter/pull/16081
